### DISCUSSION TITLE
Inline Wasm in the build

### DISCUSSION THREAD

**Discussion by nikgraf** [automerge/automerge#651] [Ideas]
My team recently built a WebAssembly library compiled from Rust and while doing the examples we ran into several issues with bundlers and wasm imports. Internally in the past we sometimes base64 encoded wasm and inlined it to avoid these problems. Luckily rollup has this functionality as well and so we switched over to that. Example is here: https://github.com/serenity-kit/opaque/blob/main/bin/build.js

Downside is that the total file size is larger (37% according to https://stackoverflow.com/a/11402374/837709), but when gzipped in our case we saw around 10% increase.
In addition we need to export a `isReady` promise so people can wait for it to load.
`libsodium-wrappers` is doing the same or similar by exposing a `ready` promise and making sure users don't have to be worried about wasm. 

Benefit is zero bundler configuration, which is often quite tough.

**Comment by jjallaire** [comment]
There are some performance tradeoffs to this approach, esp. with larger WASM payloads. Here's a related discussion: https://github.com/denoland/deno/issues/11918. I agree that doing this would ease integration for a lot people, but I'd still leave open the possibility of using a bundler.

It looks like there is also a Stage 2 proposal to add native WebAssembly module loading to JS (https://github.com/WebAssembly/esm-integration/blob/main/proposals/esm-integration/README.md) which once supported by browsers, node, deno, etc. would provide a much nicer solution to the problem.
